zcash_client_backend/src/scanning.rs

```diff
@@ -345,42 +327,91 @@ pub(crate) fn scan_block_with_runner<
     prior_block_metadata: Option<&BlockMetadata>,
     mut batch_runner: Option<&mut TaggedBatchRunner<P, K::Scope, T>>,
 ) -> Result<ScannedBlock<K::Nf>, ScanError> {
-    if let Some(scan_error) = check_continuity(&block, prior_block_metadata) {
+    if let Some(scan_error) = check_hash_continuity(&block, prior_block_metadata) {
         return Err(scan_error);
     }
 
     let cur_height = block.height();
     let cur_hash = block.hash();
 
-    // It's possible to make progress without a Sapling tree position if we don't have any Sapling
-    // notes in the block, since we only use the position for constructing nullifiers for our own
-    // received notes. Thus, we allow it to be optional here, and only produce an error if we try
-    // to use it. `block.sapling_commitment_tree_size` is expected to be correct as of the end of
-    // the block, and we can't have a note of ours in a block with no outputs so treating the zero
-    // default value from the protobuf as `None` is always correct.
-    let mut sapling_commitment_tree_size = block
-        .chain_metadata
-        .as_ref()
-        .and_then(|m| {
-            if m.sapling_commitment_tree_size == 0 {
-                None
-            } else {
-                let block_note_count: u32 = block
-                    .vtx
-                    .iter()
-                    .map(|tx| {
-                        u32::try_from(tx.outputs.len()).expect("output count cannot exceed a u32")
-                    })
-                    .sum();
-                Some(m.sapling_commitment_tree_size - block_note_count)
-            }
+    let initial_sapling_tree_size = prior_block_metadata.and_then(|m| m.sapling_tree_size());
+    // If we're below Sapling activation, or Sapling activation is not set, the
+    // tree size is zero
+    let nu_sapling_tree_size =
+        params
+            .activation_height(NetworkUpgrade::Sapling)
+            .and_then(|sapling_activation| {
+                if cur_height < sapling_activation {
+                    Some(0)
+                } else {
+                    None
+                }
+            });
+    let mut sapling_commitment_tree_size = initial_sapling_tree_size
+        .or_else(|| {
+            block
+                .chain_metadata
+                .as_ref()
+                .and_then(|m| {
+                    if m.sapling_commitment_tree_size == 0 {
+                        nu_sapling_tree_size
```

Comment:
```suggestion
                        None
```
This is equivalent because of the `.or(nu_sapling_tree_size)`. But then `nu_sapling_tree` size is only used in one place, so it can be replaced by its definition (see suggestion below), which I think will be clearer.

Reply:
Thinking aloud: 0 is the default value for unsigned integer fields in protobuf, but before Sapling activation the size of the tree is known to be zero. However, now that I think about it, it doesn't make sense to set the tree size to Some(0) in `nu_sapling_tree_size` because what we actually want is for the code that ultimately uses this value to return an error in any case where the tree size is unknown _or_ where it would be invalid to attempt to manipulate the tree. So the `nu_` variables don't need to exist; their uses can all be replaced by `None` and then the calling code will have to change so that it doesn't prematurely force the observation of a tree size.

Reply:
So, I went down the road of implementation described above, and realized that this won't work correctly without the tree sizes in `ScannedBlock` being non-optional. The whole check against the block metadata having a commitment tree size of zero is wrong; if we're given block metadata at all, we have to treat its given commitment tree sizes as correct because we need the case where we're after activation but the note commitment tree size is actually zero to be treated correctly.